src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py

```diff
@@ -714,15 +696,36 @@ def __call__(
                     if callback is not None and i % callback_steps == 0:
                         callback(i, t, latents)
 
-        # 9. Post-processing
-        image = self.decode_latents(latents)
+        if output_type == "latent":
+            image = latents
+            has_nsfw_concept = None
+
+        elif output_type == "pt":
+            # 8. Post-processing
+            image = self.decode_latents(latents)
+            has_nsfw_concept = None
+
+        elif output_type == "np":
+            # 8. Post-processing
+            image = self.decode_latents(latents)
 
-        # 10. Run safety checker
-        image, has_nsfw_concept = self.run_safety_checker(image, device, prompt_embeds.dtype)
+            # 9. Run safety checker
+            image = self.vae_feature_extractor.decode(image, output_type="np")
+            image, has_nsfw_concept = self.run_safety_checker(image, device, prompt_embeds.dtype)
 
-        # 11. Convert to PIL
-        if output_type == "pil":
-            image = self.numpy_to_pil(image)
+        elif output_type == "pil":
+            # 8. Post-processing
+            image = self.decode_latents(latents)
+
+            # 9. Run safety checker
+            image = self.vae_feature_extractor.decode(image, output_type="np")
+            image, has_nsfw_concept = self.run_safety_checker(image, device, prompt_embeds.dtype)
+
+            # 10. Convert to PIL
+            image = self.vae_feature_extractor.numpy_to_pil(image)
+
+        else:
+            raise ValueError(f"Unsupported output_type {output_type} ")
```

Comment:
I would maybe suggest to do the following logic here:

```py
if output_type == "latent":
    return (image, False)

image = self.decode_latents(latents)

image, has_nsfw_concept = self.run_safety_checker(image, device, prompt_embeds.dtype)

image = self.vae_feature_extractor.decode(image, output_type=output_type)
```

Reply:
Think this is all we need. Note that this logic implies that we have some PT -> Numpy conversion in the safety checker but that's fine